(#Closes #2592 and 2582(?)) Routine has a _symbol attribute.

LonelyCat124 opened the pull request:
@sergisiso @arporter I have a very initial implementation of the changes I feel are needed for the `Routine` node specifically to support this change. I'm a bit concerned about what I needed to do w.r.t to handling the creation of a `Routine` that doesn't have a parent and adding a parent in later (which is probably something reasonable to be able to do), as I needed to override `_parent` as both a property and setter.

If either you have time to have a look and let me know how you feel about this implementation before I move much further, and any suggestions/alternatives would be appreciated.

arporter commented:
Thanks @LonelyCat124, I think what you're missing is use of the `attach` and `detach` mechanism. See https://psyclone-dev.readthedocs.io/en/latest/psyir.html#the-parent-child-relationship and possibly the section after it on dynamic updates.

sergisiso commented:
@LonelyCat124 I had a look at this, I think overriding the _parent setter is a good possible alternative and it should not cause problems with the ChildrenList automatic parent handling. (the other alternative is using `Container._update_signal` but this will be executed any time there is ANY update below a container, which could quickly be very expensive, so I prefer to explore your `_parent.setter`). I belive you also need to delete the symbol when parent is disconnected (the `_parent.setter` is given a parent different than `_parent_node`, before discarding the old `_parent_node`, it needs something like: `_parent_node.symbol_table.remove(self._symbol)`?)

What I am not sure is about what you did inside the name.setter "if not self._symbol" branch. I think if the Scope already have a symbol with the given name it could be because another routine or import already has it, and should fail. Maybe for new Routines should just set `self._symbol = RoutineSymbol("name", NoType())` in the `__init__`, before calling the supper constructor, would that work?

sergisiso commented:
> I think if the Scope already have a symbol with the given name it could be because another routine or import already has it.

I see that one problem with this is that the fparser2reader forward-declares routine symbols (with the _process_routine_symbols calls). Because Fortran can have a Call to a routine that is declared below itself. Maybe we could modify the frontend to put empty Routines in a first pass (of the caller of _process_routine_symbols), so the symbols exist, and then populate them on a second pass.

LonelyCat124 commented:
I've merged master in (and fixed the conflict) and I now get some failing tests that aren't clear as to why they're failling for me.

One such example is `test_generate_container` in `fparser2_container_test`.

The code seems to assume that the `container` generated from `generate_container` shouldn't have any children, wheras now it does (the `subroutine` it `contains`).  Was there a reason this doesn't have a child on master? I guess creating the Routine is now adding it to the parent container, but if you use the full fparser chain it would already do that...

arporter commented:
`generate_container()`'s days are numbered - #1115. Perhaps they've now run out?

LonelyCat124 commented:
I'm not sure - what was their original use-cases? If they're used in any scripts or any other part of PSyclone its not great to remove them, I can just fix the tests (there are also other failing tests I'm yet to examine).

arporter commented:
> Edit: grepping in PSyclone only shows the tests using it, so it depends on any other uses externally.

In that case I would remove it. I'm not aware of any external uses of it. @JulienRemy do you use `fparser2reader.generate_container()`?

JulienRemy commented:
We've never used the attribute I implemented in the closed PR nor `fparser2reader.generate_container()`. No problem with doing this differently as far as I know!

LonelyCat124 commented:
Moving this into Routine looks to have some complex side effects w.r.t the DSL frontends.

One example is in GOcean where simply initialising a GOcean PSyIR tree is problematic. From what I understand, it looks like we first get a standard PSyIR tree, and then the `raise_psyir_2_gocean_kern_trans` takes standard PSyIR nodes and replaces them with the GOcean specific ones (e.g. GOceanKern).

The problem now is that since the Routine has added its' symbol to the `GOceanContainer`, there are also `Call` nodes to the `RoutineSymbol`, so when the transformation attempts to replace the Routine with the GOceanKern, the `Routine` attempts to remove its own symbol from the Container's symbol table, which fails in `_validate_remove_routinesymbol`.

I'm not really sure how to resolve this issue - any advice would be appreciated.

arporter commented:
That does sound difficult. Could we perhaps overload the `replace_with()` method of `Node` for a `Routine` and do something special in there?

LonelyCat124 commented:
Yeah - I'm just not sure what the "special" thing should be. I can't decide if we should force removal of the symbol (somehow) in this case, but I'm not sure what the result of that would be, nor if we have that functionality.

I did wonder about not adding the symbol to the parent if its a GOceanContainer or an LFRicContainer (or perhaps if its not either specifically a `Container` or `FileContainer` with type(..) != comparisons), but I'm not sure if that would break lowering of Kerns to Routines, since I'm not sure how they're lowered. It appears like it might work, but I'm not sure when the children of a GOceanContainer are lowered based on its `lower_to_language_level` call, which doesn't appear to loop through its children and lower them.

LonelyCat124 commented:
There's also some issues with outputs it seems like due to copying the tree.

When copying the Container, it keeps the symbol in the new containers symbol table (i think? Its not clear that this always happens since not every test fails...), but for one of the inline trans tests (that i'm currently looking at) it fails here:
```
psyir/backend/visitor.py:182: in __call__
    tree_copy = node.root.copy()
psyir/nodes/node.py:1602: in copy
    new_instance._refine_copy(self)
psyir/nodes/scoping_node.py:102: in _refine_copy
    super(ScopingNode, self)._refine_copy(other)
psyir/nodes/node.py:1585: in _refine_copy
    self.children.extend([child.copy() for child in other.children])
psyir/nodes/node.py:261: in extend
    self._set_parent_link(item)
psyir/nodes/node.py:185: in _set_parent_link
    node._parent = self._node_reference
psyir/nodes/routine.py:194: in _parent
    parent.symbol_table.add(self._symbol)
```
because `sub` is already in the SymbolTable, which I can only assume was due to copying in, but I'm not sure how I avoid this.

LonelyCat124 commented:
TODO:

- When replacing a Routine with a Routine - pass the symbol from the original routine to the new one (and adjust name accordingly?).

- ScopingNode check for copied Routines and remove symbols corresponding to them.

LonelyCat124 commented:
I fixed the above things, just roaming through failing tests and fixing those that I understand that are caused by these changes.

I've hit one that I don't understand how I can have changed the behaviour of the code.
In `frontend/fparser2_test.py` there is `test_deferred_array_size` which now fails because:
`KeyError: "Symbol table already contains a symbol with name 'n'.`

I don't think I modified the behaviour of process_declarations so I'm confused how this passed before - fparser2.py adds the `n` symbol to the symbol table as an unresolved integer type, and then it fails when it comes across the actual symbol `n`'s declaration.

LonelyCat124 commented:
I also seem to have some failing tests in inline_trans because the order of `Routine`s inside a `Container` has changed (and doesn't match the input code). I'll try to work out what can have changed to cause this first as it may fix a number of tests.

LonelyCat124 commented:
Ok it again looks like not an easy fix. The test is simple:
```
    code = (
        "module test_mod\n"
        "contains\n"
        "  subroutine run_it()\n"
        "    integer :: i\n"
        "    real :: a(10)\n"
        "    do i=1,10\n"
        "      call sub(a, i)\n"
        "    end do\n"
        "  end subroutine run_it\n"
        "  subroutine sub(x, ivar)\n"
        "    real, intent(inout), dimension(10) :: x\n"
        "    integer, intent(in) :: ivar\n"
        "    integer :: i\n"
        "    do i = 1, 10\n"
        "      x(i) = 2.0*ivar\n"
        "    end do\n"
        "  end subroutine sub\n"
        "end module test_mod\n")
```

During fparser, we create empty `Routine` for run_it and sub and add them to the test_mod `Container`. The implementation we decided on for filling them was we then add the internal ASTs, and then detach them (to be reconnected to their parent in `process_nodes` as is currently done).
Unfortunately for this example case this doesn't work as expected since `run_it` contains a call to `sub`, when calling `routine.detach()` for `sub` it returns a `ValueError` as its not safe to detach the node as there is a reference to its symbol.

An alternative solution to this would be to not detach `Routine` nodes as current and return None from the `_subroutine_handler`, however there is a bad case where a `Routine` should result in a `CodeBlock`, but its symbol is called elsewhere in the file, which I expect would mean the `Routine` fails to detach and adding a `CodeBlock` of this `Routine` would result in two implementations of the `Routine` in the tree, where one is potentially partly filled with the code and the other is a full implementation from a CodeBlock.

I wonder if I should force routine.detach() to work even if its symbol can't be removed, but I'm not sure what downsides this might have? Any ideas @sergisiso?

LonelyCat124 commented:
The changes break the test for lines 404-410 in `psyir/tools/call_tree_utils.py`, I'm not quite sure I understand this code, but I think it becomes impossible to fail to find a (resolved) Routine in a module now so I'm not sure this error can happen? Would need @hiker to check that section I think. `test_get_non_local_read_write_info_errors` also fails - I assume for a similar reason but I'm not sure.


Still one test failing in fparser2.py. `test_deferred_array_size ` is one, as previously discussed. I also broke some error code in `_process_decln` which is no longer reachable as the `own_routine_symbol` tag is no more in existence - I instead did a workaround by searching for the ancestor Routine and checking for return_symbol which seems to resolve those failing tests.

My remaining concern is that I broke something in adjoint @arporter. The test `test_lfric_adjoint::test_generate_lfric_adjoint_multi_routine` fails when outputting during `symbol_table.deep_copy` at this section:
```
for routine in symbol.routines:
    new_routines.append((new_st.lookup(routine.symbol.name), routine.from_container))
symbol.routines = new_routines
```
What seems to be happening is there are some symbols that refer to a "routine" `test_code_r4` which has (I assume) been adjointed into `adj_test_code_r4`, and so when it tries to lookup the `test_code_r4` symbol that it needs to sort out the `GenericInterfaceSymbol` it fails because its not been copied into the new symbol table. Its not there because its no longer present in the old one now either. This all happens during `_refine_copy` of the `fortran_writer` - it might be best if we can have a discussion about this to work out whats going wrong/what I've changed that has broken expectations.


All the other remaining tests are around kernel_module_inline_trans - I've left those alone for now - I think they're all happening due to symbols already being present in symbol tables but I need to dive into why this happens for some (but not all) tests.

LonelyCat124 commented:
Ok, the KernelModuleInlineTrans bugs were super easy to fix (it was adding the symbol manually before adding the Routine as a child which of course was breaking). The only remaining issues are the ones in the previous post now. I will merge master and make this ready for a first review despite the last few failing tests.

One for @arporter or @sergisiso I think.

hiker commented:
> The changes break the test for lines 404-410 in `psyir/tools/call_tree_utils.py`, I'm not quite sure I understand this code, but I think it becomes impossible to fail to find a (resolved) Routine in a module now so I'm not sure this error can happen? Would need @hiker to check that section I think. `test_get_non_local_read_write_info_errors` also fails - I assume for a similar reason but I'm not sure.
> 

Sorry, noticed this a bit too late, I'll try to have a look tomorrow.

One issue (that I recently found) that might(!!), trigger that code is issue #2657(?)

LonelyCat124 commented:
Check return_symbol is setup.

hiker commented:
> > The changes break the test for lines 404-410 in `psyir/tools/call_tree_utils.py`, I'm not quite sure I understand this code, but I think it becomes impossible to fail to find a (resolved) Routine in a module now so I'm not sure this error can happen? Would need @hiker to check that section I think. `test_get_non_local_read_write_info_errors` also fails - I assume for a similar reason but I'm not sure.
> 
> Sorry, noticed this a bit too late, I'll try to have a look tomorrow.
> 
> One issue (that I recently found) that might(!!), trigger that code is issue #2657(?)

@LonelyCat124 , the test originally removes the psyir for the surboutine from the container, but does not remove the symbol from the symbol table. Your code now removes the symbol, meaning instead of returning the name for a subroutine that does not exist anymore, it now returns an empty list (which is indeed better of course). 

My feeling is (don't have enough time to dig into that) that #2657 could trigger the
same code:
```
module my_api
    use some_mod, only: my_proc
end module my_api

...
   use my_api, only: my_proc
   call my_proc()
```
When trying to get the PSyIR for my_proc, it will find the symbol, but no PSyIR.

My suggestion would be to either try the above code and replace this test, or put a #TODO 2657 with an explanation there (and remove the lines that print the warning in call_tree_utils), so when I (probably my task since I need it) get to 2657, I can re-evaluate this.

hiker commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -160,6 +187,41 @@ def node_str(self, colour=True):
+        return self._parent_node
+
+    @_parent.setter
+    def _parent(self, parent):
```
While I am not asked for a review, while studying the problem of `call_tree_utils`, I noticed this code. I think this functionality should be implemented as an explicit function with a descriptive name. I saw the code:
```
node._parent = None
```
There is no way to know that this will actually modify a symbol table (and really annoying while debugging). While I am aware that my attitude towards properties is different from the rest of the PSyclone team, there is imho just too much happening as a side affect of a simple assignment.

LonelyCat124 replied:
I see your point here with the side effects, but the challenge is the `Routine` is in charge of controlling where its `RoutineSymbol` lives now already.
This `_parent` set (or rather the ChildrenList setting up parent links) also has a number of side effects that are somewhat not obvious anyway depending on the tree I think since the `update_signal` happens anyway I think from the ChildrenList.

LonelyCat124 replied:
I think `_parent` should almost always only be set from functions (e.g. `detach()`).

hiker replied:
What about overwriting `detach` in `Routine` instead? That is much easier to see (e.g. if you debug, immediately when single stepping into `detach` you would notice that something special is happening).

From my point of view, the typical pattern is that `_x` is the attribute, and then `x` a property 'function'. This breaks this convention.

LonelyCat124 replied:
I think unfortunately there is an issue because `_parent` is almost always actually set from `ChildrenList.set_parent_link` or other similar functions (e.g. when using addchild). This also meant I had to override `_parent`.

The specific choice of using `_parent` as the setter here is to keep it marked as protected as setting the parent of a Node is not something intended to be widely accessed, and this is required to match the rest of the code. Its a slightly awkward implementation, but I don't think there's another way around it.

sergisiso replied:
@hiker  I share your concerns about properties doing too much magic under-the-hood. But I think the deciding factor is if what they do is unexpected, or the opposite - not doing it would be more unexpected. Also, it's good that the public API, including properties, are atomic operations (wrt not leaving the tree in an inconsistent state, with a mandatory fix expected from the user).

In the case of _parent, it is part of a double-linked relationship with children. And before we had it automatically updated it was broken after some transformations because the responsability was on the transformation writter to "fix" the doble-linked list. This has not been problem anymore since its manages by the node itself.

This PR tries to solve a similar problem in that a Rountine must always have its implementation (node) as well as its declaration (parent scope routinesymbol). But we also inadvertedly break this some times as currently users are responsible to know that if they remove, rename, replace a routine, they must update the associated routinesymbol in the symboltable, maybe restore the "own_routine_symbol" tag, not leave the routinesymbols hanging, maybe update interfaces, have a way for the detached standalone routine to find its own symbol... (and the issues that @LonelyCat124  found are because we don't even remember to do/review all this things ourselfs) 

This PR started as an experiment to see if we can get rid of the finicky "own_routine_symbol" but also to see if we can make attaching/detaching routines be more atomic in that they don't levave the tree inconsistent forcing the user to remember to fix the symbols/references. Doing it in "detach" is not sufficient because any list operation in children (e.g. `container.children = [routine1]` could remove a previous existing routine). (Unfortunately) when debugging at least you would need to know that you need to step into this list operation, but I take your point that we should put the effects in descriptive methods.

LonelyCat124 commented:
Thanks @hiker - I also looked through it with Andy and Sergi and we realised that while the `Routine` can't be there, there could be a `CodeBlock` with the `RoutineSymbol` still present, so the code is still reachable. I've adjusted the tests to simulate that case and we added additional error messages to handle the update to the old cases.

sergisiso commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -64,19 +73,29 @@ class Routine(Schedule, CommentableMixin):
 
-    def __init__(self, name, is_program=False, **kwargs):
+    def __init__(self, name, is_program=False, symbol=None,
+                 symbol_tag=None, **kwargs):
```
If this accepts pre-existing symbols, I would prefer this to not deal with tags. I think doing the symbol in a separate step is reasonable for this cases, but this all follows from the assumption that pre-existing symbols are fine. Is this the case?

LonelyCat124 replied:
Ok - so the assumption at the moment is that pre-existing symbols are fine, but they're not expected to be in the parent's symbol table already (since the assumption is the Routine is meant to be in control of that).

Part of the reason the Routine needs to know about the symbol_tag is to ensure its carried over if the symbol is detached and then reattached, e.g. during a tree copy. I'm not sure if the tags are required for any correctness in the `FortranWriter`, but if someone wanted to copy a tree and do operations on it, the tags would need to be copied and the only way this can happen at the moment is if the Routine knows about the tag itself.

Does this make sense @sergisiso

sergisiso replied:
>  Part of the reason the Routine needs to know about the symbol_tag is to ensure its carried over if the symbol is detached and then reattached, e.g. during a tree copy

A tree copy should do a symbol_table.deep_copy for every scope, this will copy the symbols and add the same tags to the newly equivalent instances of the symbols.

This implementation had several issues that @arporter fixed in a PR that we merged yesterday, could you bring this branch to master and see if it is still a problem?

LonelyCat124 replied:
Yes - but this doesn't work with this new Routine implementation (see Container._refine_copy) - in fact this is the main challenge here.

Since copy makes a copy of the tree, it will make a copy of the Routines and a copy of the symbols but they're not linked to each other. The _refine_copy removes the copied symbols to be replaced by the Routine parent setter so the tag is lost unless the Routine has it - fixing the `deep_copy` in symbol table won't fix this issue.

sergisiso commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -64,19 +73,29 @@ class Routine(Schedule, CommentableMixin):
                             f"but got '{type(is_program).__name__}'")
         self._is_program = is_program
 
+        # Symbol is set-up by the name setter property.
+        # name property is taken from the symbol.
+        # We do this last to avoid any other errors resulting in a
+        # symbol in a SymbolTable that we shouldn't have.
+        self._has_set_name = False
+        self.name = name
```
It is unclear to me what takes precedence, the name or the symbol.name? But regardless, I think it would be much simpler if we could just have one (and fix #1200) so there is no conflict.

I guess both cases are useful sometime, but maybe we could use the fact that we have 2 constructors (init and create) and have them use a different argument.

Usually use try to have the init leaner and we use it for top-down creation, so I suggest removing the "name" here and make the symbol mandatory. But then, we usually have logic to create inner object in the "create" method and use use it for bottom-up tree creation (which means the parent symbol table may even not exist). So I suggest removing symbol from there and make the method always create the RoutineSymbol for the given name. 

This would require adding/deleting the .create from current uses (including specialisations e.g.[GO/Dyn]KernelSchedule) but I believe it will make the implementation easier and the API more predictable. What do you think?

LonelyCat124 replied:
I'm a bit confused on this comment - the routine has no `_name`, the `name` property and setter are entirely based upon the symbol object creation or passed in.

The setup in Routine initialisation are as follows:
1. There was no `symbol` argument, so we create a `RoutineSymbol` with the `name` argument's name.
2. We take the symbol (either created or passed as an argument) and add it to the parent's symbol table
3. We set `self._has_set_name = True` so future calls to `name` will take the alternative code path where the symbol's name is update.

The one thing I should probably do is check that the `symbol.name` and `name` are the same if `symbol` is provided, or I guess **make name optional and require either that `name` or `symbol` is provided which I think is probably best?** @sergisiso 

This is essentially doing what you want, but it does it all from inside the `name` setter instead of it relying on modifying other sections of the code (and means we don't need to modify the rest of the code). Its a little opaque perhaps but I don't think it should affect any subclasses unless they override `name`.

sergisiso replied:
To clarify, my comment was only about the duplication in the input arguments: "name" and "symbol".name. As you say one solution would be to "require either that name or symbol", but the other solution is to provide just symbol in the init (and discard all the _has_set_name). And then move the string "name" option to the "create" method (and remove symbol from there), in this case the create method would first create the RoutineSymbol with the given name, and then call init with it.

This would match the style of other node types where we keep the init as lean as possible, but we have create arguments that accept strings and get transformed to the underlying needed nodes with the logic inside create.

What I don't know is if this will help simplify the current _has_set_name/ name setter implementation. If possible I would prefer not to have a "routine.name setter", and rename routines by renaming their symbol with the symbol_table.rename_symbol method explicitly, as this is the same we need to do for other symbols.

(I am aware that this will require changes in the current routine usage code, because some `Routine("name")` will need to be updated to `Routine.create("name")` and `routine.name = "new_name"` will need to be updated to `routine.scope.symbol_table.rename_symbol(routine.symbol, "new_name")` )

Would this help make the implementation easier, or are there problems that I am not seeing?

sergisiso replied:
btw, if they are not attached to an scope (as you mention in another comment), could we do:
`routine.symbol = RoutineSymbol("new_name")`?
which is the same API that we have for References

LonelyCat124 replied:
> I'm happy to require Routine to require the symbol input argument and change the name to be in create, that seems ok.

This will have a lot of side effects in its subclasses and probably make the create function less clean (it will need kwargs since subclasses have extra subclasses) but i think thats ok.

sergisiso replied:
Before you do the work, what are the kwargs/subclasses that we will need?

sergisiso reviewed (changes requested):
@LonelyCat124 I just made a few top-level comments as I want to see if my assumptions are right, and if things can be simplified by using a different behaviour for the create method before I continue the review.

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/raise_psyir_2_gocean_kern_trans.py:
```diff
@@ -241,7 +241,8 @@ def apply(self, node, options=None):
         gokernsched = GOKernelSchedule(metadata.procedure_name,
-                                       symbol_table=gotable.detach())
+                                       symbol_table=gotable.detach(),
+                                       symbol=routine._symbol)
```
I added the property so we can do this now.

LonelyCat124 commented on src/psyclone/domain/lfric/lfric_kern.py:
```diff
@@ -699,7 +699,8 @@ class creates the PSyIR schedule on first invocation which is
         ksched = KernelSchedule(sched.name,
-                                symbol_table=sched.symbol_table.detach())
+                                symbol_table=sched.symbol_table.detach(),
+                                symbol=sched._symbol)
```
Done

LonelyCat124 commented on src/psyclone/psyir/symbols/generic_interface_symbol.py:
```diff
@@ -46,7 +46,9 @@ class GenericInterfaceSymbol(RoutineSymbol):
-    :type routines: list[:py:class:`psyclone.psyir.symbols.RoutineSymbol`]
+    :type routines: list[tuple[
+                             :py:class:`psyclone.psyir.symbols.RoutineSymbol`,
+                             bool]
```
Fixed

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1097,11 +1094,18 @@ def _insert_opencl_init_routine(self, node):
         subroutine = container.children[0]
-        # Rename subroutine
-        subroutine.name = subroutine_name
+        sym_tab = subroutine.symbol_table.deep_copy()
+        generated_code = subroutine.pop_all_children()
+
+        real_subroutine = Routine(subroutine_symbol.name,
+                                  symbol=subroutine_symbol,
+                                  symbol_table=sym_tab,
+                                  symbol_tag="ocl_init_routine")
+        for nodes in generated_code:
+            real_subroutine.addchild(nodes)
```
No, this is because the Routine doesn't expect its symbol to be attached to a symbol table when its created, so this is the workaround at the moment (I think the creation of this Routine object from Fortran source in the middle of the psyir was somewhat unexpected for me so I didn't have a good plan on how to handle it).

sergisiso replied:
> No, this is because the Routine doesn't expect its symbol to be attached to a symbol table when its created

But in this case we don't create it. The routine comes from the "psyir_from_source" above, so it is already attached and has a parent scope with a symbol table.

LonelyCat124 replied:
Yes - but because the Routine can't expect the symbol to be attached to the symbol table the problem is the `tag` setting. The old code this wasn't an issue as it was adding the RoutineSymbol with `add_symbol` and providing a tag. This code is that same functionality but with the new system.

LonelyCat124 replied:
There is an alternative option here (but I think its worse/handles more changes worse).

We can now instead do
```
subroutine.detach()                                                                                                                              node.addchild(subroutine)
sym = symtab.lookup(subroutine.name)
symtab.remove(sym)
symtab.add(sym, "ocl_init_routine")
```

The issue that could happen here is that the subroutine has no idea about whether its meant to have a tag, so if for some reason its detached and reattached it won't keep the tag. I'm not sure if this is a concern or not but I prefer the original implementation just in case.

sergisiso replied:
What if we reverted the routine creation to the original:
```
# Create the symbol for the routine and add it to the symbol table.
subroutine_symbol = symtab.new_symbol("psy_init",
                                       symbol_type=RoutineSymbol,
                                       tag="ocl_init_routine")
```
(note I removed the .name)

Add a routine.symbol property that only works when there is no _parent. As then we are free to change the symbol without breaking scopes.

and here just did:
```
subroutine.detach()
subroutine.symbol = subroutine_symbol
```

The tag will be kept, the replace_with tags will also be kept with my other suggestion? Would this allow us to remove the symbol_tag from the routine constructor?

LonelyCat124 replied:
This would work here (though we still don't have a .symbol access, its still using _symbol for now unless you think its ok to allow this more widely). I'm just slightly wary of losing tag information if someone did 
```
routine.detach()
routine.attach(new_parent)
```

If the routine doesn't know anything about the tag and the user was expecting the tag to be carried over from the old parent to the new parent (for whatever reason) we'd lose it

sergisiso replied:
I am not sure I agree that this should keep the tag. The tag is a property of the scope, not the symbol. If we move a datasybol from one scope to another we also lose the tag. The method that intent on keeping the tag, e.g. copy or merge already take take to transfer it.

(If this behavior needs to be changed this is a separate issue and consistent for all symbols)

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -932,9 +932,8 @@ def _insert_ocl_arg_setter_routine(node, kernel):
-        # Create the new Routine and RoutineSymbol
-        node.symbol_table.add(RoutineSymbol(sub_name), tag=sub_name)
-        argsetter = Routine(sub_name)
+        # Create the new Routine
+        argsetter = Routine(sub_name, symbol_tag=sub_name)
```
I fixed this now by just doing:
`sub_name = node.symbol_table.next_available_name(sub_name)`

LonelyCat124 replied:
Yes - but not if that symbol is already in the symbol table (because there was no current method to distinguish that behaviour and attempting to place a Routine somewhere where the symbol for it already exists).

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1044,10 +1043,8 @@ def _insert_opencl_init_routine(self, node):
-        subroutine_name = symtab.new_symbol("psy_init",
-                                            symbol_type=RoutineSymbol,
-                                            tag="ocl_init_routine").name
+        # Create the symbol for the routine.
+        subroutine_symbol = RoutineSymbol("psy_init")
```
I fixed this now by just doing:
`sub_name = node.symbol_table.next_available_name(sub_name)`

LonelyCat124 commented on src/psyclone/domain/common/transformations/kernel_module_inline_trans.py:
```diff
@@ -278,12 +278,9 @@ def apply(self, node, options=None):
-                    name, interface=DefaultModuleInterface()
-            ))
-            # 2) Insert the relevant code into the tree.
+            # If it doesn't exist already, module-inline the subroutine by
+            # inserting the relevant code into the tree. The Routine now
+            # automatically places its symbol into the symbol table.
```
I removed the additions I made to the comment, but I'd like to leave the original comment section.

LonelyCat124 commented:
@sergisiso I addressed the comments (some with changes, some with discussion) and merged Andy's symbol table deep copy changes into the master too.

sergisiso commented on src/psyclone/psyir/nodes/scoping_node.py:
```diff
@@ -119,11 +125,57 @@ def _refine_copy(self, other):
+            try:
+                if isinstance(self._symbol_table.lookup(node.name),
+                              RoutineSymbol):
+                    self._symbol_table.remove(
+                            self._symbol_table.lookup(node.name))
+            except KeyError:
+                pass
+            except ValueError:
+                # If remove fails for the RoutineSymbol then we have to force
+                # the removal of the symbol.
+                symbol = self._symbol_table.lookup(node.name)
+                norm_name = self._symbol_table._normalize(symbol.name)
+                self._symbol_table._symbols.pop(norm_name)
```
If we have to force it anyway, it could be better to replace:
`self._symbol_table.remove(self._symbol_table.lookup(node.name))`
with:
```
# Replace it manually to avoid checking if there are still references to this symbol,
# those references will be also replaced below.
symbol = self._symbol_table.lookup(node.name)
norm_name = self._symbol_table._normalize(symbol.name)
self._symbol_table._symbols.pop(norm_name)
```
And remove the "except ValueError", since it's the expected path anyway.

LonelyCat124 replied:
Yeah - I was initially concerned whether we had to fix any `Call` nodes to point to this symbol, but that has never happened (I had an `assert False` to check for this) so I think that this is safe. Done.

sergisiso commented on src/psyclone/psyir/nodes/scoping_node.py:
```diff
@@ -119,11 +125,57 @@ def _refine_copy(self, other):
+        super(ScopingNode, self)._refine_copy(other)
+        # pylint: disable=protected-access
+
+        # Update of children references to point to the equivalent symbols in
+        # the new symbol table attached to self.
+        # TODO #1377 Unfortunately Loop nodes currently store the associated
+        # loop variable in a `_variable` property rather than as a child so we
+        # must handle those separately. Also, in the LFRic API a Loop does not
+        # initially have the `_variable` property set which means that calling
+        # the `variable` getter causes an error (because it checks the
+        # internal-consistency of the Loop node). We therefore have to check
+        # the value of the 'private' `_variable` for now.
+        # We have to import Loop here to avoid a circular dependency.
+        # pylint: disable=import-outside-toplevel
+        from psyclone.psyir.nodes.loop import Loop
+        for node in self.walk((Reference, Loop)):
+            if isinstance(node, Reference):
+                if node.symbol in other.symbol_table.symbols:
+                    node.symbol = self.symbol_table.lookup(node.symbol.name)
+            if isinstance(node, Loop) and node._variable:
+                if node.variable in other.symbol_table.symbols:
+                    node.variable = self.symbol_table.lookup(
+                        node.variable.name)
```
Is this not the same that the `replace_symbols_using` is doing just below?

LonelyCat124 replied:
I think perhaps I broke `replace_symbols_using` and all of this code somehow when attempting to fix conflicts - I'll check master.

LonelyCat124 replied:
Nvm I misunderstood how that function worked so the code above can be removed.

sergisiso reviewed (changes requested):
@LonelyCat124 See some more comments mainly regarding routine and scope_node changes. I have not resolved some previous comments yet as I want to understand this part better first.

LonelyCat124 commented on src/psyclone/]:
```diff
@@ -0,0 +1,2299 @@
+# pylint: disable=too-many-lines
```
Yeah my bad, i thought I did this sorry.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -96,8 +110,8 @@ def __eq__(self, other):
     @classmethod
-    def create(cls, name, symbol_table, children, is_program=False,
-               return_symbol_name=None):
+    def create(cls, name, symbol_table=None, children=None, is_program=False,
+               symbol=None, return_symbol_name=None, **kwargs):
```
Yeah I can add a default argument to these constructors and avoid it being passed into create - the only concern would be if some other code wanted to do this but I feel thats unlikely.

LonelyCat124 replied:
Done.

sergisiso replied:
Thanks for updating InvokeSchedule constructors, they better now.

What about the first comment about removing "symbol" from create? is it possible?

LonelyCat124 replied:
Yeah we can remove the symbol create, it was only being used in a test to test an exception so I'll remove it.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -51,9 +53,14 @@ class Routine(Schedule, CommentableMixin):
-    :param bool is_program: whether this Routine represents the entry point \
-                            into a program (e.g. Fortran Program or C main()).
+    :param symbol: the Symbol used to represent this Routine in its Container.
+                   If it is not supplied a RoutineSymbol will be created with
+                   default arguments instead.
```
The reason we're not fine with it if its already there is that we have no way of determining why its already there - e.g. is there another Routine in this scope using that symbol that we've piggy-backed on (which would be an error). I'll add a note for this.

LonelyCat124 replied:
I've written a note to attempt to emphasize the why, and removed the out-of-date section of the docstring I think.

sergisiso replied:
> e.g. is there another Routine in this scope using that symbol that we've piggy-backed on

@LonelyCat124 but wouldn't this be another RoutineSymbol with the same name, would it be fine to let it pass if it is the same object ("is" comparison)?

sergisiso replied:
Ah you mean like attaching the same routine node (or a copy of it) 2 times? Since we have the scope, could a scope.walk(Routine) tell us if thats the case and then complain, but be fine if not found?

LonelyCat124 replied:
I think mostly I'm wary of it to avoid mistakes, e.g. if you have this code

```
subroutine my_sub()
...
end subroutine my_sub
```

and you did
```
fc = my_sub.ancestor(FileContainer)
routine_symbol = fc.symbol_table.lookup("my_sub")
# Meant to do .replace but made a mistake
replacement = Routine(routine_symbol, parent=fc)
```

I think this should fail, because otherwise PSyclone may make a Fortran output like:
```
subroutine my_sub()
...
end subroutine

subroutine my_sub()
...
end subroutine
```

which I think is unlikely to be intended output.

I guess your solution could work (though it would need to be `scope.walk(Routine, CodeBlock)` and be able to check the CodeBlock name as well (I'm not sure how this works). I can change it to that if you think it'd be appropriate.

LonelyCat124 replied:
Added support for this, just need to update the documentation.

LonelyCat124 replied:
Also refactored the code slightly.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -300,11 +331,65 @@ def _refine_copy(self, other):
+                "Routine class should be the same as the Routine being "
+                "replaced.")
+        # We want to not use the childrenList update information.
+        node._symbol_in_table = self._symbol_in_table
```
`_symbol_in_table` is persistent and keeps track of whether we were able to remove the symbol when detaching the `Routine` from its parent. When we use replace then `node._symbol_in_table = True` is probably accurate (since `self` has to have a parent to be able to be replaced), but it needs to be updated in case it is then `detach`ed somehow later.

I guess alternatively we could adjust the `_parent` setter to do `self._symbol_in_table = True` if the `ValueError` is hit when removing it from its parent and that would solve this issue without a private access, but the `_symbol_in_table` does need to be correct for correct behaviour.

LonelyCat124 replied:
Added a comment for this, but happy to adjust this if you think it would be better.

sergisiso replied:
If I understand this correctly you use this only for the replace_with case, because in that case the remove could fail (there may be calls to it), but we know it is fine because we are re-adding a Routine with the same symbol (`self._symbol != node._symbol`) check. If this is the case, could you call it

`self._routine_being_replaced = False`

and maybe in the setter just do
```
if self._routine_being_replaced == True:
   # If this comes from a replace_with, we don't want to remove the routine when the setter
   # is given the None value nor add it again when the new parent is given. The second call
   # with a not None parent marks the end of the replace with
   if parent is not None:
       self._routine_being_replaced = False
   self._parent_node = parent  # This line can also be gone if using the "update_parent_symbol_tables"
else:
   # logic without the _routine_being_replaced
```
Would this work?

LonelyCat124 replied:
No thats not it, its primarily for the `remove` case as we don't fail to `remove` the Routine itself, only its symbol may continue to exist in the symbol table it was removed from. So if its reattached to a parent, its used to determine whether it needs to create a copy of its symbol to place in its new parent (since the old symbol is already in a symbol table elsewhere). Failing to remove the routine on `.detach()` didn't feel like a good choice.

LonelyCat124 commented on src/psyclone/psyir/nodes/scoping_node.py:
```diff
@@ -71,6 +71,12 @@ def __init__(self, children=None, parent=None, symbol_table=None):
             # Create a new symbol table attached to this scope
             self._symbol_table = self._symbol_table_class(self)
 
+        # As Routines add their own symbol into their parent's symbol
+        # table its necessary to ensure that the _symbol_table exists
+        # before we add the children to the Node (which happens in the
+        # super constructor).
```
I think this comment is explaining the lines above - the previous implementation I think did the super call before the symbol table attachment (which broke for Routines). I'll move it above

LonelyCat124 replied:
Done

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -64,14 +71,21 @@ class Routine(Schedule, CommentableMixin):
+        self._symbol = symbol
+        self._symbol_tag = symbol_tag
+        self._symbol_in_table = False
+        self._parent_node = None
```
Yes - since `_parent` is a function and not a variable the actual `_parent` is stored in `_parent_node`.

LonelyCat124 replied:
Comment added

sergisiso replied:
I believe you can still access the parent variable with super()._parent, but maybe it will be clearer if instead of overloading the properties we just did:
```
  def __setattr__(self, name, value):
      if name == '_parent':
          self.update_parent_symbol_table(value)
      super().__setattr__(name, value)

  def update_parent_symbol_tables(self, new_parent):
      """ Remove self._symbol from self._parent.symbol_table and add it to new_parent.symbol_table """
```

Alternatively, we could just added this method to the places where "_parent" is meant to be set which are all at Node  (in `set_parent_link`, `del_parent_link`, `__init__`, and `_refine_copy`) e.g.:
```
        node.update_parent_symbol_tables(self._node_reference)
        node._parent = self._node_reference
``` 
It is not as automatic but, we would still have `routine._parent =` if we would want to do something nasty in a particular situation or test and really force the reference change without the rest of the logic.

What do you think?

LonelyCat124 replied:
You can find the parent with `._parent` but I think whenever you do `._parent = X` it would call the setter (so the property and setter reference ._parent_node as the actual storage location) - at least this is my understanding of how it can work.

I will try the first implementation you suggested and unless anything breaks we can use that :)

I don't think we can just override those functions in node as `ChildrenList` does `._parent` setting directly.

LonelyCat124 replied:
This change did break some things in copying and initialization, so I had to add a little code to `update_p;arent_symbol_table` to handle that too.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -300,11 +331,65 @@ def _refine_copy(self, other):
+                "replaced.")
+        # We want to not use the childrenList update information.
+        node._symbol_in_table = self._symbol_in_table
+        # Just calling node._parent_node.children.__setitem__ attempt
+        # to update the _parent of the replacements. Due to the nature
+        # of the replace_with and symbol functionality on Routines, this
+        # will not work. Instead we use the list.__setitem__ method to
+        # update this directly.
```
This is a comment w.r.t an old implementation - the `_symbol_in_table` handles this instead and enables us to just use the super now instead.

LonelyCat124 replied:
This is removed now.

LonelyCat124 commented:
@sergisiso Can pass back to you to have another look. I'm happy to adjust to have the code allow an attached RoutineSymbol as long as no Routine/CodeBlock exists in the scope with the same name if you think that is appropriate.

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -322,10 +326,14 @@ class GOInvokeSchedule(InvokeSchedule):
     _text_name = "GOInvokeSchedule"
 
-    def __init__(self, name, alg_calls, reserved_names=None, parent=None):
+    def __init__(self, name, alg_calls=None, reserved_names=None, parent=None,
```
s/name/symbol/ ?

LonelyCat124 replied:
Fixed.

sergisiso commented on src/psyclone/tests/psyir/backend/fortran_unsupported_declns_test.py:
```diff
@@ -94,16 +94,16 @@ def test_fw_unsupported_interface_decln(tmpdir, fortran_writer):
-    # Add remaining structures so that we can generate compilable code
-    container.symbol_table.add(RoutineSymbol("eos1d"))
-    container.symbol_table.add(RoutineSymbol("eos2d"))
+    # Remove the my_sub RoutineSymbol before we readd it
+    container.symbol_table.remove(container.symbol_table.lookup("my_sub"))
```
There is a typo in "readd", but before that, could we just not move the:
`container.addchild(Routine.create("my_sub", SymbolTable(), []))`
above in place of the 
`container.symbol_table.add(RoutineSymbol("my_sub", visibility=Symbol.Visibility.PUBLIC))`
so then we don't need to remove it?

LonelyCat124 replied:
Yeah thats neater, done.

sergisiso commented on src/psyclone/tests/psyir/backend/sir_test.py:
```diff
@@ -225,9 +225,12 @@ def test_sirwriter_node_1():
+    class Unsupported(ScopingNode):
         '''A PSyIR node that will not be supported by the SIR writer but
         accepts any children inside.'''
+        def __init__(self):
+            super().__init__()
+
```
The empty constructor shouldn't be needed.

And why did you need to subclass ScopingNode here?

LonelyCat124 replied:
We're adding a Routine to an `Unsupported`, so the Routine's parent needs to have a symbol table (i.e. be a ScopingNode or similar).

sergisiso reviewed (changes requested):
That's another nice improvement @LonelyCat124 , I just closed some previous comments that I think are now resolved and made a light review, before I continue can you give a shot at this:

> I'm happy to adjust to have the code allow an attached RoutineSymbol as long as no Routine/CodeBlock exists in the scope with the same name if you think that is appropriate.

The issue I have as the things currently are is that some code feels out of place. For instance, you had to manually do the inner steps to copy routines in a few places, and replicating the routine copy logic sparsely around the code, but all of them are missing the upadate_symbols_with (which to be fair it was introduced after you started the PR, but I think it shows that this logic is not well encapsulated). I have similar worries about Routine having to know that symbol tags exist at all, it doesn't feel it should be it's responsibility.

Looking that Routine/CodeBlock exists in the scope will have other complexities but I am wondering it those will be better encapsulated inside the Routine and not around the code.

LonelyCat124 commented on src/psyclone/psyad/adjoint_visitor.py:
```diff
@@ -442,6 +442,22 @@ def _copy_and_process(self, node):
             if isinstance(result, Node):
                 result = [result]
             node_copy.children.extend(result)
+
+        # Update any GenericInterfaceSymbols to the use the new routine
+        # symbols created during copying.
+        sym_tab = node_copy.symbol_table
+        for symbol in node_copy.symbol_table.get_symbols():
+            symbol_obj = sym_tab.lookup(symbol)
+            if isinstance(symbol_obj,
+                          GenericInterfaceSymbol):
+                new_routines = []
+                for rsym, mod_pro in symbol_obj.routines:
+                    new_routines.append((
+                            sym_tab.lookup(rsym.name),
+                            mod_pro
+                    ))
+                symbol_obj.routines = new_routines
+
```
I'm not sure, this was before Andy's changes I think, I'll try commenting it out and see.

LonelyCat124 replied:
No its not, I think the issue is adjoint modifys the name of the symbols in the symbol table after copying, but this doesn't propagate into the `GenericInterfaceSymbol`. Its in the visitor for the routines  I think, and this was the only place I could find to do this.

LonelyCat124 commented:
@sergisiso I implemented that change and I added a comment above showing how we could remove some of the "copying" of things - I think that implementation is worse due to how tags currently work, but maybe there's a better solution to tags somehow.

sergisiso reviewed (changes requested):
@LonelyCat124 I closed a few comments but also proposed 3 extra changes in previous conversations that we could explore to improve the remaining issues.

LonelyCat124 commented:
@sergisiso I responded to the comments, I'll upstream the changes I've made when the tests have run - maybe the last 2 points are best to have a discussion once you're back from leave next week?

LonelyCat124 commented:
Remove tag from symbol.
Enable public setter (and accesor) for .symbol
Ensure that replace_with doesn't lose tag infomation.

LonelyCat124 commented:
@sergisiso `test_named_invoke_name_clash` in the `test_psyGen.py` is now failing and I'm not sure why/what is meant to do the symbol renaming here.
The code now generates:
`SUBROUTINE invoke_a(invoke_a, b, istp,` instead of `SUBROUTINE invoke_a(invoke_a_1, b, istp,`, but I'm not even sure what I'm generating or the code route since I've never really used these older parts of the code.

sergisiso commented:
The name of invoke arguments is given at: `src/psyclone/psyGen.py(2201)_complete_init()`

At that line if I print the symtab with the debugger on master I get:
```
(Pdb) print(symtab)
Symbol Table of InvokeSchedule 'invoke_a':
------------------------------------------
DataSymbol:
  cell: DataSymbol<Scalar<INTEGER, i_def: DataSymbol<Scalar<INTEGER, UNDEFINED>, Import(container='constants_mod')>>, Automatic>
RoutineSymbol:
  invoke_a: RoutineSymbol<NoType, pure=unknown, elemental=unknown>
Symbol:
  STENCIL_1DX: Symbol<Automatic>
  STENCIL_1DY: Symbol<Automatic>
...
```

with this branch only:
```
(Pdb) print(symtab)
Symbol Table of InvokeSchedule 'invoke_a':
------------------------------------------
DataSymbol:
  cell: DataSymbol<Scalar<INTEGER, i_def: DataSymbol<Scalar<INTEGER, UNDEFINED>, Import(container='constants_mod')>>, Automatic>
Symbol:
  STENCIL_1DX: Symbol<Automatic>
  STENCIL_1DY: Symbol<Automatic>
...
```
I am not sure why this happened with the last merge, I will keep exploring.

LonelyCat124 commented:
Hmm, i'm not sure why its not in the parent now, I'll try to investigate when I can.

sergisiso commented:
Digging a bit more is because this called from inside the `schedule_class` in psyGen.py line 413:
```
        # create the schedule (Routine sub-class). Routines use a
        # symbol input argument.
        schedule_symbol = RoutineSymbol(self._name)
        self._schedule = schedule_class(schedule_symbol,
                                        alg_invocation.kcalls,
                                        reserved_names, parent=container)

        # Add the new Schedule to the top-level PSy Container
        if container:
            container.addchild(self._schedule)
```
But the RoutineSymbol is not added into the container in the addchild step below.

I think before you where adding the symbol in the parent contructor argument `parent=container` , and now it is only added in the proper addchild?

sergisiso commented:
But maybe its a more general problem than this particular bit of code. Which is that for an unconnected Routine we will not be able to find its symbol in the symbol_table (e.g. for check for name clashes). Would it made sense to ALSO add the same RoutineSymbol object in the Routine symbol_table? This will mean that detaching it will still make a self-consitent subtree, and we will find the name of the symbol when checking the symbol table.

LonelyCat124 commented:
> But maybe its a more general problem than this particular bit of code. Which is that for an unconnected Routine we will not be able to find its symbol in the symbol_table (e.g. for check for name clashes). Would it made sense to ALSO add the same RoutineSymbol object in the Routine symbol_table? This will mean that detaching it will still make a self-consitent subtree, and we will find the name of the symbol when checking the symbol table.

Do we need to be careful for this case? Is it not valid to have `subroutine a( a, ...)`? The issue adding the routine symbol back into its own symbol table is that it overlaps with the return symbol for `function` routines.

sergisiso commented:
> Do we need to be careful for this case? Is it not valid to have subroutine a( a, ...)? The issue adding the routine symbol back into its own symbol table is that it overlaps with the return symbol for function routines.

If there already exist it is fine, we don't need to add the routine symbol. But if not, it could still be a recursive subroutine (which in later standards the recursive keyword is optional). For instance:
```
subroutine a(b, c)
    call a(b,c)
end subroutine
```
should be a self-consistent tree even when detached from a container. So it may be good to add it in the inner scope if it's not already there (shadowed by a function return value)?

LonelyCat124 commented:
I have a partial implementation of Routine adding its symbol to its own symbol table, which is then removed if the return symbol shares a name (in fparser2.py). There are a bunch of other things now failing but I'll try to work through them

LonelyCat124 commented:
@sergisiso I think this is ready for another look now - I added in the changes we discussed as far as I recall so hopefully we're getting close now.

sergisiso reviewed (changes requested):
@LonelyCat124 I agree this is close now, and I am happy about how it turn out. I closed all previous comment. Most of the remaining ones are clean up and adding some more specific test in routine_test.py

LonelyCat124 commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1141,10 +1142,9 @@ def test_named_invoke_name_clash(tmpdir):
+    assert ("SUBROUTINE invoke_a(invoke_a, b, istp, rdt, d, e, ascalar, "
             "f, c, g, qr)") in gen
-    assert "TYPE(field_type), intent(in) :: invoke_a_1" in gen
+    assert "TYPE(field_type), intent(in) :: invoke_a" in gen
```
Ah damn, this has broken again since the `RoutineSymbol` isn't in the Routine again, as we remove the Routine's symbol from itself if it has a parent.
I'm not sure I have a good solution to prevent this behaviour - the previous solution of adding the RoutineSymbol into the routine broke for return symbols. 
Any ideas @sergisiso - we could add something into symbol table to prevent the same name for a symbol in a `Routine` but I'm not sure how we allow it for return symbols - it could also happen at lowering where we rename any symbol that shares a name with the Routine if its not the return symbol?

sergisiso replied:
This gets solved (this change can be reverted and no other tests fail) by removing the: `if self.parent is None and self.symbol_table:` line in `Routine.__init__` (the self.symbol_table is not needed because this is guaranteed by the super init). Something like this:

```
-        # If the routine is detached we try to
-        # add the symbol into it's symbol table unless the symbol table
-        # already contains a symbol with that name (e.g. for a Function).
-        if self.parent is None and self.symbol_table:
-            try:
-                self.symbol_table.lookup(self._symbol.name)
-            except KeyError:
-                self.symbol_table.add(self._symbol)
+        # Add the symbol into the routine itself, unless the symbol table
+        # already contains a symbol with that name (e.g. for a Function)
+        if self._symbol.name not in self.symbol_table:
+            self.symbol_table.add(self._symbol)
```

This means that we are always adding itself in the subroutine, in addition to the parent module when this is connected. I also checked that functions (with and without result clause) end up having the right symbols (wihout result the routinesymbols is replaced with a datasymbol). Do you think this is a feasible solution? Or are there drawbacks?

I am unsure if we could also simplify `update_parent_symbol_table` after this change.

LonelyCat124 replied:
My fear with this was that it broke functions, but since they're ok I'm happy to make this change.

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -932,9 +932,11 @@ def _insert_ocl_arg_setter_routine(node, kernel):
-        node.symbol_table.add(RoutineSymbol(sub_name), tag=sub_name)
-        argsetter = Routine(sub_name)
+        # Create the new Routine
+        sub_name = node.symbol_table.next_available_name(sub_name)
+        sub_symbol = RoutineSymbol(sub_name)
+        node.symbol_table.add(sub_symbol, tag=kernel.name + "_set_args")
```
Reverted (n.b. the tag is still different to what you put but it needs to be to be findable).

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1018,7 +1020,7 @@ def _insert_ocl_arg_setter_routine(node, kernel):
         node.addchild(argsetter)
 
-        return node.symbol_table.lookup_with_tag(sub_name)
+        return node.symbol_table.lookup_with_tag(kernel.name + "_set_args")
```
I don't agree, the tag needs to be computable at entry to this function so should be `kernel.name + "_set_args" and not the result from next_available_name (since this is not consistent across multiple calls).

sergisiso replied:
I agree, it was my mistake, resolving issue

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1097,11 +1097,17 @@ def _insert_opencl_init_routine(self, node):
-        # Rename subroutine
-        subroutine.name = subroutine_name
+        sym_tab = subroutine.symbol_table.deep_copy()
+        generated_code = subroutine.pop_all_children()
+
+        symtab.add(subroutine_symbol, tag="ocl_init_routine")
+        real_subroutine = Routine(subroutine_symbol,
+                                  symbol_table=sym_tab)
+        for nodes in generated_code:
+            real_subroutine.addchild(nodes)
 
         # Add the subroutine as child of the provided node
-        node.addchild(subroutine.detach())
+        node.addchild(real_subroutine)
```
Done

LonelyCat124 replied:
The subroutine_symbol variable is needed I think as the symbol needs adding to the symbol table manually to make sure we have the tag (and I'd rather have the variable personally than have to look it up to give it to the Routine).

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1128,10 +1134,8 @@ def _insert_initialise_grid_buffers(node):
-        subroutine_name = symtab.new_symbol("initialise_grid_device_buffers",
-                                            symbol_type=RoutineSymbol,
-                                            tag="ocl_init_grid_buffers").name
+        # Create the symbol for the routine.
+        subroutine_symbol = RoutineSymbol("initialise_grid_device_buffers")
```
Done

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1222,10 +1232,8 @@ def _insert_write_grid_buffers(self, node):
-        subroutine_name = symtab.new_symbol("write_grid_buffers",
-                                            symbol_type=RoutineSymbol,
-                                            tag="ocl_write_grid_buffers").name
+        # Create the symbol for the routine.
+        subroutine_symbol = RoutineSymbol("write_grid_buffers")
```
Done

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1304,10 +1318,8 @@ def _insert_ocl_read_from_device_function(self, node):
-        subroutine_name = symtab.new_symbol("read_from_device",
-                                            symbol_type=RoutineSymbol,
-                                            tag="ocl_read_func").name
+        # Create the symbol for the routine.
```
Done

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1397,10 +1414,8 @@ def _insert_ocl_write_to_device_function(self, node):
-        subroutine_name = symtab.new_symbol("write_to_device",
-                                            symbol_type=RoutineSymbol,
-                                            tag="ocl_write_func").name
+        # Create the symbol for the routine.
```
Done

LonelyCat124 commented on src/psyclone/domain/gocean/transformations/gocean_opencl_trans.py:
```diff
@@ -1490,10 +1511,8 @@ def _insert_ocl_initialise_buffer(node):
-        subroutine_name = symtab.new_symbol("initialise_device_buffer",
-                                            symbol_type=RoutineSymbol,
-                                            tag="ocl_init_buffer_func").name
+        # Create the symbol for the routine .
```
Done

LonelyCat124 commented on src/psyclone/domain/lfric/lfric_invoke.py:
```diff
@@ -77,7 +77,8 @@ def __init__(self, alg_invocation, idx, invokes):
         # pylint: disable=import-outside-toplevel
         from psyclone.domain.lfric import LFRicInvokeSchedule
-        self._schedule = LFRicInvokeSchedule('name', None)  # for pyreverse
+        # for pyreverse
+        self._schedule = LFRicInvokeSchedule.create('name')
```
Removed

LonelyCat124 commented on src/psyclone/gocean1p0.py:
```diff
@@ -217,7 +217,8 @@ class GOInvoke(Invoke):
     '''
     def __init__(self, alg_invocation, idx, invokes):
-        self._schedule = GOInvokeSchedule('name', None)  # for pyreverse
+        # for pyreverse
+        self._schedule = GOInvokeSchedule.create('name')
```
A bunch of tests fail without this - I'm just going to remove the comment.

LonelyCat124 commented on src/psyclone/gocean1p0.py:
```diff
@@ -2153,9 +2161,7 @@ def _read_from_device_routine(self):
-        subroutine_name = symtab.new_symbol(
-            "read_from_device", symbol_type=RoutineSymbol,
-            tag="openacc_read_func").name
+        subroutine_symbol = RoutineSymbol("read_from_device")
```
Fixed - can't use new_symbol since it doesn't return it and i don't want to also have to then lookup the symbol.

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -2052,18 +2023,9 @@ def _process_decln(self, scope, symbol_table, decl, visibility_map=None,
-                        # the same tag
-                        tag = "own_routine_symbol"
-                        raise KeyError
+                    # RoutineSymbols will never be shadowed as FParser will
+                    # always have a container as a parent of a Routine.
+                    # Check is the symbol is unresolved.
```
Removed

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -5347,128 +5307,166 @@ def _subroutine_handler(self, node, parent):
+        routine = None
+        # The Routine may have been forward declared in
+        # _process_routine_symbol, and so may already exist in the
+        # PSyIR as a child of parent with no subtree. If so, we find the
```
Yeah

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -5347,128 +5307,166 @@ def _subroutine_handler(self, node, parent):
+        # _process_routine_symbol, and so may already exist in the
+        # PSyIR as a child of parent with no subtree. If so, we find the
+        # Routine object and fill it with the Routine internals.
+        # If we find a Routine object in this way we also need to detach it
+        # from its parent so it can be added to its parent in _create_child.
```
Yeah I think this was outdated - I've removed that part of the comment.

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -5347,128 +5307,166 @@ def _subroutine_handler(self, node, parent):
+
+                # Ensure that we have an explicit declaration for the symbol
+                # returned by the function.
+                keep_tag = None
```
Yeah its only set here so removed.

LonelyCat124 commented on src/psyclone/psyir/nodes/node.py:
```diff
@@ -1747,6 +1749,17 @@ def replace_symbols_using(self, table):
+        :param new_parent: The new parent of this node.
+        :type new_parent: :py:class:`psyclone.psyir.nodes.ScopingNode`
+
+        This base implementation does nothing.
```
Done

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -64,19 +73,33 @@ class Routine(Schedule, CommentableMixin):
+                            f"'{type(symbol).__name__}'")
+        self._parent = None
+        self._symbol = symbol
+#        self._symbol_in_table = False
```
Removed all commented out references to symbol_in_table now.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -194,53 +320,20 @@ def name(self, new_name):
                 try:
-                    rsymbol = self.parent.symbol_table.lookup(new_name).copy()
+                    sym = self.symbol_table.lookup(symbol.name)
+                    if sym is self._symbol:
```
So this is the case where we have a detached routine, so sym is either the RoutineSymbol or another symbol that shadows its name (i.e. the returnsymbol).

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -282,15 +398,12 @@ def return_symbol(self, value):
             raise TypeError(f"Routine return-symbol should be a DataSymbol "
                             f"but found '{type(value).__name__}'.")
-        if value not in self.symbol_table.datasymbols:
+        if (value not in self.symbol_table.datasymbols):
```
Yeah - removed. Not sure why flake8 didn't flag that.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -300,11 +413,62 @@ def _refine_copy(self, other):
         :type other: :py:class:`psyclone.psyir.node.Routine`
 
         '''
+        # Removing the parent_node knowledge from the default copy method
+        # before calling the super, else in Node we set self._parent = None
+        # and attempt to remove this symbol from a symbol table that doesn't
+        # contain it.
+        self._parent_node = None
+        self._symbol = other._symbol.copy()
```
Yeah, removed.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -300,11 +413,62 @@ def _refine_copy(self, other):
+                "The symbol of argument node in method replace_with in the "
+                "Routine class should be the same as the Routine being "
+                "replaced.")
+        # Set the symbol_in_table status so when using the _parent setter
+        # we don't attempt to overwrite the symbol, as its already in the
+        # symbol table.
+#        node._symbol_in_table = self._symbol_in_table
```
Done

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -160,6 +188,104 @@ def node_str(self, colour=True):
+        :raises GenerationError: if a Codeblock representing a routine with
+                                 the same name already exists in the scope.
+        '''
+        # Need to check the _parent property exists as the initial set of the
+        # _parent causes this to fail otherwise
```
Yeah comment is out of date, removed.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -160,6 +188,104 @@ def node_str(self, colour=True):
+        # Need to check the _parent property exists as the initial set of the
+        # _parent causes this to fail otherwise
+        if self._parent is not None:
+            try:
+                # Need a check that this Routine's symbol is in the current
+                # _parent symbol table, as otherwise this breaks during
+                # copying.
+                if (self._parent.symbol_table.lookup(self.name)
+                        is self._symbol):
+                    self._parent.symbol_table.remove(self._symbol)
+#                    self._symbol_in_table = False
+            except ValueError:
+                pass
+            except KeyError:
+                pass
```
It wasn't originally, but since we've improved the way the refine_copy works that might be the only remaining failure case so I'll add a TODO.

LonelyCat124 commented on src/psyclone/psyir/nodes/routine.py:
```diff
@@ -160,6 +188,104 @@ def node_str(self, colour=True):
         '''
         return self.coloured_name(colour) + "[name:'" + self.name + "']"
 
+    def update_parent_symbol_table(self, new_parent):
```
Added some tests for this

LonelyCat124 commented:
@sergisiso I fixed most of the new comments - only the last one is outstanding as we need to work out a way to fix it as I'm unsure on when/how to do it.

sergisiso commented:
@LonelyCat124 I resolved most issues, I will have a think about the remaining problem this afternoon.

(PS: It seems `src/psyclone/tests/psyir/nodes/test_file.py` has been added by mistake?)

arporter commented:
Hi @LonelyCat124, don't worry about the LFRic integration test failures yet - Sergi's token for gist has expired. I've removed these uploads in #2708.

sergisiso commented:
Unfortunately the integration fails in the NEMOv4 and NEMOv5 pass-throughs before the gist uploading with:

`NVFORTRAN-S-0155-Redefinition of kiss (/home/gh_runner/NEMO/cfgs/SPITZ12_passthrough/WORK/storng.f90: 36`

Kiss is a function with a CONTAINS and other functions inside (should be a codeblock?), you can see it here: https://forge.nemo-ocean.eu/nemo/nemo/-/blob/main/src/OCE/STO/storng.F90

LonelyCat124 commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -5347,128 +5307,166 @@ def _subroutine_handler(self, node, parent):
+            # we still need to ensure the symbol is in the parent's symbol
+            # table. For this case the best we can do is place the symbol
+            # in the tree without a coresponding Routine.
+            sym = parent.symbol_table.lookup(routine.name)
```
Done

LonelyCat124 commented:
I'm going to rerun the integration tests now after the error from alst week. Not sure why project coverage decreases though.

arporter commented:
A drop in coverage often indicates 'indirect changes' - i.e. changes in coverage of lines that aren't directly touched in this PR. If you look at https://app.codecov.io/gh/stfc/PSyclone/pull/2596/indirect-changes you'll see you have a few.

LonelyCat124 commented:
If this now works with coverage I'll run integration tests once more as I removed what I believe was unreachable code now with this branch, but I want to double check integration tests don't break.

LonelyCat124 commented:
@sergisiso 
I'm still having NEMOv4 fail with this error:
```
nvfortran -c -o storng.o -I/home/gh_runner/NEMO/cfgs/SPITZ12_passthrough/BLD/inc -i4 -Mr8 -g -O3 -Mflushz -Mfma -Minline -fastsse -Mcray=pointer -Mvect=prefetch,tile,fuse -Mpre -mp=gpu -gpu=managed -I/home/gh_runner/actions-runner/_work/PSyclone-mirror/PSyclone-mirror/lib/profiling/nvidia/ -I/home/aporter/NEMO/xios_trunk/inc -I/not/defiled/build/lib/mct -I/not/defiled/build/lib/psmile.MPI1 -I/apps/packages/netcdf_fortran/4.6.1-nvidia-hpcsdk/include -c /home/gh_runner/NEMO/cfgs/SPITZ12_passthrough/WORK/storng.f90
NVFORTRAN-S-0155-Redefinition of kiss (/home/gh_runner/NEMO/cfgs/SPITZ12_passthrough/WORK/storng.f90: 28)
```

We tested this file I thought and it was fine - I'll try to have another look at this tomorrow.

LonelyCat124 commented:
@sergisiso I found the issue, the output contains this:
```
  subroutine kiss()


  end subroutine kiss
  ! PSyclone CodeBlock (unsupported code) reason:
  !  - PSyclone doesn't yet support 'Contains' inside a Subroutine or Function
  FUNCTION kiss()
```

What seems to happen is the fparser2.py:5311 doesn't remove the fake Subroutine.

I have a test that checks this subroutine is removed and will fix this up now.

LonelyCat124 commented:
@sergisiso This should be ready for another look now - it made it past the failing part of integration so should be ok now.

sergisiso reviewed (approved):
@LonelyCat124 All the comments have been addressed, all the tests and integration test pass and the documentations builds. This is ready to merge

Reading and grepping the branch I found a bunch of `symbol_table.lookup_with_tag('own_routine_symbol')` and `routine.symbol_table.lookup(routine.name)` or similar instances, and I replaced all of these with routine.symbol. This has not broken any test, but I will run a final integration test just in case.

sergisiso commented:
Integration test passed again, there was just some flake8 issues that I introduced myself and I have now fixed.
